### PR TITLE
Delta Exporter: Azure Support

### LINKED ISSUE #7296: Catalog Export Hooks for Unity to support Azure

### PR DESCRIPTION
**Pull request by N-o-Z**
Closes #7296
### Background

Add support for delta catalog export with Azure

### Testing Details

Added new esti test for delta catalog export with AWS and Azure falvors

### Breaking Change?

No

### CODE REVIEW

File: esti/export_hooks_files/delta/azure/_lakefs_actions/delta_export.yaml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        local delta_exporter = require("lakefs/catalogexport/delta_exporter")
+        
+        local table_descriptors_path = "_lakefs_tables"
+        local sc = azure.client(args.azure.storage_account, args.azure.access_key)
+        local delta_client = formats.delta_client(args.lakefs.access_key_id, args.lakefs.secret_access_key)
+        local delta_table_locations = delta_exporter.export_delta_log(action, args.table_names, sc.put_object, delta_client, table_descriptors_path)
+
+        for t, loc in pairs(delta_table_locations) do
+          print("Delta Lake exported table \"" .. t .. "\"'s location: " .. loc .. "\n")
+        end
```
Comment (itaiad200) [inline esti/export_hooks_files/delta/azure/_lakefs_actions/delta_export.yaml:21]: Same question applies here

File: esti/export_hooks_files/delta/azure/_lakefs_actions/delta_export.yaml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+      args:
+        azure:
+          storage_account: "{{ .AzureStorageAccount }}"
```
Comment (itaiad200) [inline esti/export_hooks_files/delta/azure/_lakefs_actions/delta_export.yaml:24]: Is this a secret? Consider just putting this as plaintext here if it isn't

File: esti/catalog_export_test.go [hunk 3]
```diff
@@ -383,3 +395,108 @@ func TestAWSCatalogExport(t *testing.T) {
...
+		return nil
+	})
+	require.NoError(t, err)
+
+	headCommit := uploadAndCommitObjects(t, ctx, repo, mainBranch, map[string]string{
+		"_lakefs_actions/delta_export.yaml": renderTplFileAsStr(t, testData, tmplDir, fmt.Sprintf("%s/_lakefs_actions/delta_export.yaml", blockstore)),
+	})
+
+	runs := waitForListRepositoryRunsLen(ctx, t, repo, headCommit.Id, 1)
+	require.Equal(t, "completed", runs.Results[0].Status)
```
Comment (itaiad200) [inline esti/catalog_export_test.go:499]: Consider checking the run result name or other identifier to make sure you're seeing the correct result

**Review by itaiad200 — APPROVED** [review]
Always great to see test coverage added, thank you! I enjoyed reading this, few questions and requests for clarifications as I'm not very familiar with our exports.

File: pkg/actions/lua/storage/client.go [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+type Client interface {
+	GetObject(l *lua.State) int
+	PutObject(l *lua.State) int
+	DeleteObject(l *lua.State) int
+	ListObjects(l *lua.State) int
+	DeleteRecursive(l *lua.State) int
+}
+
+func InitStorageClient(l *lua.State, client Client) {
```
Comment (Isan-Rivkin) [inline pkg/actions/lua/storage/client.go:13]: IMO we don't need this abstraction, I fear it's too early and make things more complex. 

To make it easier to use with exporter like `export_delta_log` you can use the function param [write_object](https://github.com/treeverse/lakeFS/blob/master/pkg/actions/lua/lakefs/catalogexport/delta_exporter.lua#L45) and pass a wrapper function  for azure storage client `write_object: function(bucket, key, data)` it doesn't need to be on the SDK level. 

1. GCS is not included here
2. In S3/Azure the interface has 2 missing functions  (delete_recursive, list_objects missing)
3. `PutObject(l *lua.State) int` is not not matching in both cases (referring to implementation with comment `// Skipping first argument as it is the host arg (bucket which is irrelevant in Azure)` in azure that you added.). 
4. Those SDK's are not meant to be used by export hooks only, our users should be able to use those SDK independently for their own needs, the combination of them here doesn't seem friendly to user experience.

**Review by Isan-Rivkin — CHANGES_REQUESTED** [review]
Hey, that's a big PR so it'll take some time but I added the most meaningful things that I found so far.

File: pkg/actions/lua/storage/azure/client.go [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+// ListObjects Should be implemented when needed. There are nuances between HNS and BlobStorage which requires understanding the
+// Actual use case before implementing the solution
+func (c *Client) ListObjects(l *lua.State) int {
+	lua.Errorf(l, "Not implemented")
+	panic("unreachable")
+}
+
+// DeleteRecursive Should be implemented when needed. There are nuances between HNS and BlobStorage which requires understanding the
+// Actual use case before implementing the solution
+func (c *Client) DeleteRecursive(l *lua.State) int {
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply pkg/actions/lua/storage/azure/client.go:114]: For conformity we created a client interface, the client objects need to implement that interface to support all the required functionalities.
This is relevant beyond the scope of the delta exporter. I think it is essential to keep the code modular.
In the future when we decide to add support for Hive export in Azure - we will need to implement it.

File: pkg/actions/lua/lakefs/catalogexport/table_extractor.lua [hunk 6]
```diff
@@ -40,7 +40,7 @@ end
 local function get_table_descriptor(client, repo_id, commit_id, logical_path)
     local code, content = client.get_object(repo_id, commit_id, logical_path)
     if code ~= 200 then
-        error("could not fetch data file: HTTP " .. tostring(code))
+        error("could not fetch data file: HTTP " .. tostring(code) ", path: " .. logical_path)
```
Comment (Isan-Rivkin) [inline pkg/actions/lua/lakefs/catalogexport/table_extractor.lua:43]: if `logical_path` is `nil` then the code will panic `attempt to concatenate a nil value`. 

```suggestion
        error("could not fetch data file: HTTP " .. tostring(code) ", path: ",  logical_path)
```

**Review by Isan-Rivkin — CHANGES_REQUESTED** [review]

File: pkg/actions/lua.go [hunk 7]
```diff
@@ -234,9 +234,10 @@ func NewLuaHook(h ActionHook, action *Action, cfg Config, e *http.Server, server
 				Config:     cfg,
 				Endpoint:   e,
 			},
-			Script:    script,
-			Args:      args,
-			collector: collector,
+			Script:        script,
+			Args:          args,
+			collector:     collector,
+			serverAddress: serverAddress,
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply pkg/actions/lua.go:240]: lakeFS server address - used to create the lakeFS client in the lua script

File: docs/howto/hooks/lua.md [hunk 8]
```diff
@@ -412,18 +412,18 @@ Parameters:
 
 A package used to export Delta Lake tables from lakeFS to an external cloud storage.
 
-### `lakefs/catalogexport/delta_exporter.export_delta_log(action, table_names, writer, delta_client, table_descriptors_path)`
+### `lakefs/catalogexport/delta_exporter.export_delta_log(action, table_def_names, write_object, delta_client, table_descriptors_path)`
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply docs/howto/hooks/lua.md:415]: This was a bug in the documentation - the code did not change

File: esti/export_hooks_files/delta/s3/_lakefs_actions/delta_export.yaml [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+        local delta_exporter = require("lakefs/catalogexport/delta_exporter")
+        
+        local table_descriptors_path = "_lakefs_tables"
+        local sc = aws.s3_client(args.aws.access_key_id, args.aws.secret_access_key, args.aws.region)
+        local delta_client = formats.delta_client(args.lakefs.access_key_id, args.lakefs.secret_access_key, args.aws.region)
+        local delta_table_locations = delta_exporter.export_delta_log(action, args.table_names, sc.put_object, delta_client, table_descriptors_path)
+
+        for t, loc in pairs(delta_table_locations) do
+          print("Delta Lake exported table \"" .. t .. "\"'s location: " .. loc .. "\n")
+        end
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply esti/export_hooks_files/delta/s3/_lakefs_actions/delta_export.yaml:21]: Not sure I understand the comment - can you explain?

File: esti/export_hooks_files/delta/azure/_lakefs_actions/delta_export.yaml (hunk 2 above)
Comment (N-o-Z) [reply esti/export_hooks_files/delta/azure/_lakefs_actions/delta_export.yaml:24]: It makes it easier to change the storage account name (in a single place) in the code and not in the yaml file

File: esti/catalog_export_test.go (hunk 3 above)
Comment (N-o-Z) [reply esti/catalog_export_test.go:499]: Done

File: docs/howto/hooks/lua.md [hunk 10]
```diff
@@ -412,18 +412,18 @@ Parameters:
...
 The function used to export Delta Lake tables.
 The return value is a table with mapping of table names to external table location (from which it is possible to query the data).
 
 Parameters:
 
 - `action`: The global action object
-- `table_names`: Delta tables name list (e.g. `{"table1", "table2"}`)
-- `writer`: A writer function with `function(bucket, key, data)` signature, used to write the exported Delta Log (e.g. `aws/s3.s3_client.put_object`)
+- `table_def_names`: Delta tables name list (e.g. `{"table1", "table2"}`)
+- `write_object`: A writer function with `function(bucket, key, data)` signature, used to write the exported Delta Log (e.g. `aws/client.s3_client.put_object` or `azure/client.client.put_object`)
```
Comment (Isan-Rivkin) [inline docs/howto/hooks/lua.md:424]: I'm not sure what `aws/client.s3_client.put_object`. 
If i understand correctly the authors intention (@Jonathan-Rosenberg ) it should be [aws/s3.put_object(bucket, key, value)](https://docs.lakefs.io/howto/hooks/lua.html#awss3put_objectbucket-key-value)

File: pkg/actions/lua/storage/azure/azure.go [hunk 11]
```diff
@@ -0,0 +1,18 @@
+package azure
+
+import (
+	"context"
+
+	"github.com/Shopify/go-lua"
+)
+
+func Open(l *lua.State, ctx context.Context) {
```
Comment (Isan-Rivkin) [inline pkg/actions/lua/storage/azure/azure.go:9]: please add docs for for the new azure lib, similar to [aws](https://docs.lakefs.io/howto/hooks/lua.html#lua-library-reference)

**Review by Isan-Rivkin — CHANGES_REQUESTED** [review]
Thank you for addressing previous comment! 
Added some comments, overall looks like a very useful addition :)

File: docs/howto/hooks/lua.md (hunk 10 above)
Comment (N-o-Z) [reply docs/howto/hooks/lua.md]: Documentation had a lot of bugs - fixed

File: pkg/actions/lua/storage/azure/azure.go (hunk 11 above)
Comment (N-o-Z) [reply pkg/actions/lua/storage/azure/azure.go:9]: Done

Comment (Isan-Rivkin) [reply pkg/actions/lua/storage/azure/azure.go:9]: Wonderful !

**Review by Isan-Rivkin — APPROVED** [review]
LGTM! Can't wait to use it!

**Pull request merged**